subsys/bluetooth/host/att.c

```diff
@@ -2866,6 +2906,70 @@ static int bt_att_accept(struct bt_conn *conn, struct bt_l2cap_chan **ch)
 BT_L2CAP_CHANNEL_DEFINE(att_fixed_chan, BT_L2CAP_CID_ATT, bt_att_accept, NULL);
 
 #if defined(CONFIG_BT_EATT)
+static k_timeout_t credit_based_connection_delay(struct bt_conn *conn)
+{
+	/*
+	 * 5.3 Vol 3, Part G, Section 5.4 L2CAP COLLISION MITIGATION
+	 * ... In this situation, the Central may retry
+	 * immediately but the Peripheral shall wait a minimum of 100 ms before retrying;
+	 * on LE connections, the Peripheral shall wait at least 2 *
+	 * (connPeripheralLatency + 1) * connInterval if that is longer.
+	 */
+
+	if (IS_ENABLED(CONFIG_BT_CENTRAL) && conn->role == BT_CONN_ROLE_CENTRAL) {
+		return K_NO_WAIT;
+	} else if (IS_ENABLED(CONFIG_BT_PERIPHERAL)) {
+		uint8_t random;
+		int err;
+
+		err = bt_rand(&random, sizeof(random));
+		if (err) {
+			random = 0;
+		}
+
+		const uint8_t rand_delay = random & 0x7; /* Small random delay for IOP */
+		const uint32_t calculated_delay =
+			2 * (conn->le.latency + 1) * BT_CONN_INTERVAL_TO_MS(conn->le.interval);
+
+		return K_MSEC(MAX(100, calculated_delay + rand_delay));
+	}
+
+	/* Must be either central or peripheral */
+	__ASSERT_NO_MSG(false);
+	CODE_UNREACHABLE;
+}
+
+static int att_schedule_eatt_connect(struct bt_conn *conn, uint8_t eatt_chans_to_connect)
+{
+	struct bt_att *att = att_get(conn);
+
+	att->eatt_chans_to_connect = eatt_chans_to_connect;
+
+	return k_work_reschedule(&att->connection_work, credit_based_connection_delay(conn));
+}
+
+static void ecred_connect_cb(struct bt_conn *conn, uint16_t result, uint8_t attempted_to_connect,
```

Comment:
I have concerns about the logic in `ecred_connect_cb`.

-
  If the outgoing connect is rejected first and then the incoming connect is rejected, the `att_schedule_eatt_connect` is called with the `attempted_to_connect - succeeded_to_connect` from the incoming request. It should be the values from the outgoing request.
 Example scenario:
  - Send L2CAP connect with 4 channels requested (because CONFIG_BT_EATT_MAX = 3).
  - Receive reject (Reason is relevant, but e.g. because the remote has allocated the channels for the next request, but not yet sent the request).
  - Receive L2CAP connect with 5 channels requested.
  - Send reject (because ATT_CHAN_MAX = CONFIG_BT_EATT_MAX).
  - Schedule a retry asking for 5 channels. (This is a bug.)

- Two outgoing EATT connects that are both rejected will trigger the retry. This is not too bad, but it's not the intention.

- This `ecred_connect_cb` handler will catch rejects for ecred connect request not related to EATT. Two rejects will trigger an unwanted eatt_connect.
  Example scenario:
   - Application tries to create a ecred connectiont to a unrelated PSM.
   - It's rejected.
   - Application tries again.
   - It's rejected again.
   - This spuriously triggers an `att_schedule_eatt_connect`.

This is not a problem with the back port. But we may want to create a new issue for this.

Reply:
@alwa-nordic this is a backport of an existing PR, https://github.com/zephyrproject-rtos/zephyr/pull/42268. Let's discuss with @hermabe whether this requires additional changes in `main` or we can merge this backport.

Reply:
I made some changes that seem to fix these issues. Please take a look: https://github.com/hermabe/zephyr/commit/dfd454f585b60d9bce30225775b18c2637cce9b6